fitgabor/trainer.py

```diff
@@ -3,32 +3,39 @@
 from torch import optim
 from tqdm import trange
 
-def trainer_fn(gabor_generator, model_neuron, 
-               epochs=20000, lr=1e-3, 
-               fixed_std=.01, 
-               save_rf_every_n_epoch=None):
-    
+
+def trainer_fn(gabor_generator, model_neuron,
+               epochs=20000, lr=1e-3,
+               fixed_std=.01,
+               save_rf_every_n_epoch=None,
+               optimizer=optim.Adam):
     gabor_generator.apply_changes()
-    
-    optimizer = optim.Adam(gabor_generator.parameters(), lr=lr)
+
+    optimizer = optimizer(gabor_generator.parameters(), lr=lr)
+
     pbar = trange(epochs, desc="Loss: {}".format(np.nan), leave=True)
     saved_rfs = []
     for epoch in pbar:
         optimizer.zero_grad()
-        
+
         # generate gabor
         gabor = gabor_generator()
-        
+
         if fixed_std is not None:
             gabor_std = gabor.std()
             gabor_std_constrained = fixed_std * gabor / gabor_std
-        
+
         loss = -model_neuron(gabor_std_constrained)
+
         loss.backward()
-        optimizer.step()
+
+        def closure():
+            return loss
```

Comment:
Shouldn't the whole thing be inside the `closure` function?

``` python
for epoch in pbar:
    def closure():
        optimizer.zero_grad()
        
        # generate gabor
        gabor = gabor_generator()

        # and other stuff
        
        loss = -model_neuron(gabor_std_constrained)
        loss.backward()
        return loss
    
    optimizer.step(closure)
```

ref: https://pytorch.org/docs/stable/optim.html#optimizer-step-closure

Reply:
I actually think this is important because the BFGS optimizer needs to evaluate the whole thing (probably mostly related to gradients) multiple times - so what the closure returns is ideally different every time that it is called by the optimizer. the way it is implemented now the return of closure is always the same.

Reply:
Yeah that is why I added it, but I missed the part about it having to return different values.

I am figuring it out now so that it can work correctly with the rest of the code.

Reply:
The fact that it returns different values is my assumption (based on intuition), but I have not tried it ever, so I might actually be wrong.

Reply:
I implemented this, however, it became really slow (from 5s per neuron to 30s) and NaNs are even more common then previously.

Reply:
hm.. ok. 
Could you push the commit? I would also like to take a look (when and if I find time)

Reply:
I pushed the commit